braingeneers/utils/memoize_s3.py

```diff
@@ -0,0 +1,145 @@
+import os
+import glob
+import boto3
+import awswrangler as wr
+from functools import partial
+from smart_open.s3 import parse_uri
+from joblib import register_store_backend, Memory
+from joblib._store_backends import StoreBackendBase, StoreBackendMixin, CacheItemInfo
+from .smart_open_braingeneers import open
+
+
+def s3_isdir(path):
+    """
+    S3 doesn't support directories, so to check whether some path "exists",
+    instead check whether it is a prefix of at least one object.
+    """
+    try:
+        next(wr.s3.list_objects(glob.escape(path), chunked=True))
+        return True
+    except StopIteration:
+        return False
+
+
+class S3StoreBackend(StoreBackendBase, StoreBackendMixin):
+    _open_item = staticmethod(open)
+
+    def _item_exists(self, location):
+        return wr.s3.does_object_exist(location) or s3_isdir(location)
+
+    def _move_item(self, src_uri, dst_uri):
+        # awswrangler only includes a fancy move/rename method that actually
+        # makes it pretty hard to just do a simple move.
+        src, dst = [parse_uri(x) for x in (src_uri, dst_uri)]
+        self.client.copy_object(
+            Bucket=dst["bucket_id"],
+            Key=dst["key_id"],
+            CopySource=f"{src['bucket_id']}/{src['key_id']}",
+        )
+        self.client.delete_object(Bucket=src["bucket_id"], Key=src["key_id"])
+
+    def create_location(self, location):
+        # Actually don't do anything. There are no locations on S3.
+        pass
+
+    def clear_location(self, location):
+        # Recursive delete.
+        wr.s3.delete_objects(glob.escape(location))
```

Comment:
I make a suggestion in a following comment, but I would enforce a cache dir prefix here and error if trying to delete outside of the cache dir.  The following suggested prefix is `s3://braingeneersdev/cache/` .  A user should not be able to accidentally clear a cache if they naively set, for example, `s3://braingeneersdev/ephys/` as their cache dir.

Reply:
That's a good point, but it's also a lot less dangerous than it looks because of the way joblib uses the cache dir you give it. A call like this:
```python
@memoize("s3://braingeneers/")
def bar(baz): ...
```
in the module `foo.py` has its actual cache files stored under `s3://braingeneers/joblib/foo/bar/`, so only things under that whole prefix get deleted when the cache is cleared.

Reply:
Perfect.  I would still check for the `s3://braingeneers/joblib` prefix then, in that case.  It's good to be a little paranoid about recursive deletion, I think.

Reply:
Yeah, fair. I'll check that the requested location actually starts with self.location + '/joblib/' and then nothing weird should happen